Resolve model bindings on route site by using mapping which pass on moment route registration.

veeeeeRySeXy opened the pull request:
#5

veeeeeRySeXy committed:
Added laravelcollective/html to composer dependencies.

veeeeeRySeXy committed:
Updated required version tymon/jwt-auth

Commit:
Resolve composer dependencies issue with dingo/api

Update saritasa/dingo-api-custom to 2.0

Commit:
Add model binding in ApiResourceRegistrar::resource

Add ApiResourceRegistrar test.
Add mocker/mockery to composer dev dependencies.

Commit:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	composer.json

Commit:
Remove ApiResourceRegistrar __call method description

Commit:
Add possibility register model bindings on route side

Add custom realization for laravel ControllerDispatcher contract
Add custom realization for laravel Registrar contract
Improve ApiResourceRegistrarTest

Commit:
Improve JWTAuthApiController

VladimirBerdnik commented on src/Api/ApiResourceRegistrar.php:
```diff
@@ -54,9 +59,18 @@ public function __construct(Router $api)
+        string $controller,
+        array $options = [],
+        string $modelClass = null,
+        string $modelName = null
```
isn't **modelName** in common cases is same as **resourceName**? What do you think, can we use resource name when model name not passed? We can use Str::singular($resourceName) to retrieve singular form

VladimirBerdnik replied:
I see that we can leave this parameter empty, but, for example, when we want to get some 'participant' of class User, should we use **'participants/{user}'** or **'participants/{participant}'**? I know that we can configure it via parameters, but which behavior should be by default?

veeeeeRySeXy replied:
When model name not passed we should use short name of given Model class. I think this behavior more clearly. How you said in common cases modelName will be as class short name

veeeeeRySeXy replied:
When we pass model class we tells, that we want to get object of this class. ModelName means that we will be found this object with this give name. Answer on your question: you can use both of these routes, but you must remember that participant should be instance of user, and when you pass participant parameter, but your controller get User $user, you should do next router call 
$api->resource('participants', 'controller', [], Participant::class, 'user')

VladimirBerdnik commented on README.md:
````diff
@@ -99,6 +99,64 @@ app('api.router')->version(config('api.version'), ['namespace' => 'Saritasa\Lara
+```
+In this case ApiController::show will receive directly parameter from url. Ex: /user/5
+
+###Route with binding on controller side
````
Here should be space after *###* and in other places too

VladimirBerdnik commented on README.md:
```diff
@@ -99,6 +99,64 @@ app('api.router')->version(config('api.version'), ['namespace' => 'Saritasa\Lara
+}
+
+// Route example
+$registrar->get('user/{user}', ApiController::class, 'show');
```
please, use plural forms for resource names even for examples (**users/{user}**)

VladimirBerdnik commented on README.md:
```diff
@@ -99,6 +99,64 @@ app('api.router')->version(config('api.version'), ['namespace' => 'Saritasa\Lara
+class ApiController
+{
+    public function show(User $user)
+	{
```
Unexpected indent. In other places too

VladimirBerdnik commented on README.md:
```diff
@@ -99,6 +99,64 @@ app('api.router')->version(config('api.version'), ['namespace' => 'Saritasa\Lara
+So if method has type hinting of class which not a parent for given in router, TypeError will be thrown.
+
+####Note:
+In 2 last cases classes which uses for model bindings must implements Illuminate\Contracts\Routing\UrlRoutable otherwise this parameters
```
*... must implement**~~s~~** ... **,** otherwise ...*

VladimirBerdnik commented on CHANGES.md:
```diff
@@ -1,5 +1,12 @@
+2.1.0
+-----
+Add model binding in ApiResourceRegistrar and WebResourceRegistrar
+Add custom realization of Illuminate\Routing\Contracts\ControllerDispatcher
```
IMHO *Add custom **implementation** of ... that allows to ...*

VladimirBerdnik commented on src/Api/ApiResourceRegistrar.php:
```diff
@@ -75,24 +89,52 @@ public function resource(string $resourceName, string $controller, array $option
     }
 
+    /**
+     * Resolves model class name. Ex: App\Models\User -> User.
```
IMHO **resolve** has other purposes, here we just *getShortClassName()* and no need to point that we can use it only for model

VladimirBerdnik commented on src/Api/ApiResourceRegistrar.php:
```diff
@@ -120,18 +169,28 @@ private function action(string $verb, string $path, string $controller, string $
 
-    private function asArray($value): array
+    /**
+     * Converts params to needed form.
```
What is **needed form** and which **params**?

VladimirBerdnik commented on src/Api/ApiResourceRegistrar.php:
```diff
@@ -75,24 +89,52 @@ public function resource(string $resourceName, string $controller, array $option
-                'uses' => $controller.'@'.$action
+            $route = $opt['route'];
+            if ($modelName) {
+                $route = str_replace('{id}', "{{$modelName}}", $opt['route']);
```
isn't better to build valid route from beginning instead of replacing them in loop? For example with `getVerbsRoutesConfig(string $attribute = '{id}'): array` ?

VladimirBerdnik commented on src/Providers/ControllersServiceProvider.php:
```diff
@@ -19,4 +24,10 @@ public function register()
         $this->app->register(RevisionsServiceProvider::class);
     }
+
+    protected function updateDefaultLaravelResolvers()
```
missed description in added function. Also IMHO it not **updates** but **replaces**

VladimirBerdnik commented on src/Router.php:
```diff
@@ -0,0 +1,76 @@
+     * Return the parameter name if it exists in the given parameters.
+     *
+     * @param string $name Router parameter to check
+     * @param array $parameters Expected parameters
```
*Available parameters*?

VladimirBerdnik commented on src/Router.php:
```diff
@@ -0,0 +1,76 @@
+use ReflectionParameter;
+
+/**
+ * Custom application router.
```
What is the difference with laravel's router?

VladimirBerdnik commented on src/Web/WebResourceRegistrar.php:
```diff
@@ -64,9 +64,18 @@ public function __construct(Router $router)
      * @param string $resourceName URI of resource
      * @param string $controller FQDN Class name of Controller, which contains action method
      * @param array $options options, passed to router on route registration
+     * @param string $modelClass Model to resolve binding
```
*Model class ...*

VladimirBerdnik commented on tests/ApiResourceRegistrarTest.php:
```diff
@@ -0,0 +1,249 @@
+use Saritasa\Laravel\Controllers\BaseController;
+
+/**
+ * Api resource registrar test
```
No tests for web resources registrar?

Commit:
Improve PHPDoc. Add unit tests on JWtAuthApiController

Add unit tests on BaseApiController
Add mockery/mockery in composer dependencies.

Commit:
Change minimal PHP version to 7.1

Commit:
Improve ApiResourceRegistrar

Commit:
Merge branch 'feature/improve-jwt-auth-api-controller'

# Conflicts:
#	composer.json

Commit:
Improve code and documentation

veeeeeRySeXy commented:
This version has not compatibility with previous. And new version required PHP >= 7.1

VladimirBerdnik commented on src/Api/ResetPasswordApiController.php:
```diff
@@ -34,31 +33,33 @@ class ResetPasswordApiController extends BaseApiController
+    protected function sendResetFailedResponse(Request $request, $message)
     {
-        return $this->json(new ErrorMessage($response));
+        return $this->json(new Message($message));
```
here should be also **trans($message)**, like in **sendResetResponse**.

VladimirBerdnik commented on src/Api/ResetPasswordApiController.php:
```diff
@@ -34,31 +33,33 @@ class ResetPasswordApiController extends BaseApiController
     protected function sendResetResponse($response)
     {
-        return $this->json(new SuccessMessage(trans($response)));
+        return $this->json(new Message(trans($response)));
```
Also, why **ErrorMessage** and **SuccessMessage** classes was replaced by **Message**?
The reason of using separate classes is that on response handler we can add 200 status code for instance of SuccessMessage and status code 400 for instance of ErrorMessage classes. Now success and failed response looks identical
